ANF application volume group support for sap-hana

abhishekkrm opened the pull request:
# PR Checklist

Check these items before submitting a PR... 

[Contribution Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md)

[Best Practice Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md)


- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

## Changelog

*
*
*

abhishekkrm commented:
Will resolve checks

abhishekkrm commented:
@AdelGhabboun  Can you kindly review this.

AdelGhabboun commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/metadata.json:
```diff
@@ -0,0 +1,9 @@
+  "type": "QuickStart",
+  "itemDisplayName": "SAP HANA Azure NetApp Files storage",
+  "description": "SAP HANA Azure NetApp Files storage",
+  "summary": "SAP HANA Azure NetApp Files storage",
```
Would be great to have more description and summary over this project!

abhishekkrm replied:
Sure

abhishekkrm replied:
Sure

AdelGhabboun commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/README.md:
```diff
@@ -0,0 +1,129 @@
+# Create Azure NetApp Files resource for SAP-HANA Storage
+
+![Azure Public Test Date](https://azurequickstartsservice.blob.core.windows.net/badges/quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/PublicLastTestDate.svg)
```
would be great to move all references to end of the Readme inside "Reference" section

abhishekkrm replied:
These are not references. I observed most of the template has the following images at the beginning. Moreover pipeline also requires to have image links.

bmoore-msft replied:
They are required at the top for samples in the repo

AdelGhabboun commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/azuredeploy.json:
```diff
@@ -0,0 +1,642 @@
+                "description": "If a Tag Value is specified, following TAG value will be added to each volume. The value will only be added if Tag Key was specified. It is recommended to add tag for HSR deployments, with tag value as \"Please enter the peering partner Volume ID\" and later updated for each volume."
+            }
+        },
+        "AzureNetAppFilesLocation": {
```
Should include default value or parameterized.

abhishekkrm replied:
Done

AdelGhabboun commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/azuredeploy.json:
```diff
@@ -0,0 +1,642 @@
+        },
+        "AzureNetAppFilesLocation": {
+            "type": "string",
+            "minLength": 1,
```
I believe this should reflect the actual min location length requirement that exist now, nothing has a length of 1. and same for max

abhishekkrm replied:
The reason why we put minLength as 1 is to ensure the location is not empty. I did check other template for min for location, other references.
https://github.com/Azure/azure-quickstart-templates/blob/5078bfd8a9ae70712311ab02d60ae9f02523fbd9/application-workloads/openshift/openshift-container-platform/nested/clusternode.json#L7

https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/resource-name-rules

Is the current restriction is minLength is 1 and maxLength is 260 ? Kindly confirm.

AdelGhabboun replied:
that works too it's just there is no region name with min length of 1.

AdelGhabboun commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/azuredeploy.json:
```diff
@@ -0,0 +1,642 @@
+                "description": "Azure NetApp Files (ANF) Location."
+            }
+        },
+        "ANFNetAppAccount": {
```
Same thing with this variable and all the other variables. need to have DefaultValue or get added to the parameter files

abhishekkrm replied:
sure, anything which needs to be provided will add them to parameter files.

AdelGhabboun reviewed (changes requested):
Submitted for few changes

abhishekkrm commented:
@gdpradhan

abhishekkrm commented:
Latest arm template requires parameters not to include ( % or \. Fixing and testing..

abhishekkrm commented:
> Latest arm template requires parameters name not to include ( % or . Fixing and testing..

Fixed the parameters Name and updated the API version @AdelGhabboun  Kindly approve again. I don't have write permission to this repo kindly help.

abhishekkrm commented:
@Brian Moore Can you help us merge this MR.

abhishekkrm commented:
The pipeline deployment is failing (2 errors during deployment in USGov and MAC) because of the pre-requisite listed in README. Some of the pre-requisite can be satisfied by creating a prereq folder and creating netapp ANF account and VNET, but for satisfying PPG requirement user must submit a request as describe here for sap hana VM pinning request.
https://docs.microsoft.com/en-gb/azure/azure-netapp-files/application-volume-group-considerations#best-practices-about-proximity-placement-groups
 
Pinning Constrain:-
**Avset pinning and PPG**: To ensure the lowest latency, the SAP HANA compute VMs need to be pinned using an Avset. A PPG needs to be created and added to the Avset and at least one VM needs to be started when calling the ARM template. Using this PPG ensures that the application volume group is able to automatically identify the Azure NetApp Files resources with the lowest latency (See also: [Best practices about proximity placement groups](https://docs.microsoft.com/en-gb/azure/azure-netapp-files/application-volume-group-considerations#best-practices-about-proximity-placement-groups))
 
This is something we cannot make it pass in the pipeline without pre-creating the PPG for the appropriate resource Group.

bmoore-msft commented:
Let's add a prereq template for the resources that are prereqs.  For PPG it sounds like that's optional.  If so, we should be able to automate testing by leaving that option out.

If it's not optional, we'll have to run a manual test.

abhishekkrm commented:
@bmoore-msft  PPG is not optional for SAP-HANA. I will add prereq template for other params.

abhishekkrm commented:
@bmoore-msft Incorporated comment. Can you kindly approve the MR ?

abhishekkrm commented:
squash all the commits.

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/prereqs/prereq.azuredeploy.json:
```diff
@@ -0,0 +1,269 @@
+            "type": "String"
+        },
+        "adminUsername": {
+            "defaultValue": "azureuser",
```
remove defaultValue

abhishekkrm replied:
done

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/prereqs/prereq.azuredeploy.json:
```diff
@@ -0,0 +1,269 @@
+                "enableIPForwarding": false
+            }
+        },
+        {
```
remove this resource, it's defined on the vnet

abhishekkrm replied:
done

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/prereqs/prereq.azuredeploy.json:
```diff
@@ -0,0 +1,269 @@
+                "privateLinkServiceNetworkPolicies": "Enabled"
+            }
+        },
+        {
```
remove this resource, it's defined on the vnet

abhishekkrm replied:
done

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/prereqs/prereq.azuredeploy.json:
```diff
@@ -0,0 +1,269 @@
+            "name": "[parameters('networkInterfaces_avgvm980_name')]",
+            "location": "[parameters('location')]",
+            "dependsOn": [
+                "[resourceId('Microsoft.Network/virtualNetworks/subnets', parameters('virtualNetworks_vnet1_name'), 'default')]"
```
make this dependent on the vnet after you remove the subnet resources

abhishekkrm replied:
done

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/prereqs/prereq.azuredeploy.json:
```diff
@@ -0,0 +1,269 @@
+                    "id": "[resourceId('Microsoft.Compute/proximityPlacementGroups', parameters('proximityPlacementGroups_avgppg_name'))]"
+                },
+                "hardwareProfile": {
+                    "vmSize": "Standard_M8ms"
```
since this is just needed for pinning, does it need to be m8?

abhishekkrm replied:
yes, otherwise we end up with ppg where we don't have ANF stamps.

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/prereqs/prereq.azuredeploy.json:
```diff
@@ -0,0 +1,269 @@
+                    },
+                    "osDisk": {
+                        "osType": "Linux",
+                        "name": "[concat(parameters('virtualMachines_avgvm_name'), '_OsDisk_1_b853d7e2ddfd432582b494d0ea69c1d2')]",
```
remove this property

abhishekkrm replied:
@bmoore-msft  can you confirm, we wan to remove the osDisk property.

abhishekkrm replied:
done

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/prereqs/prereq.azuredeploy.json:
```diff
@@ -0,0 +1,269 @@
+                        "createOption": "FromImage",
+                        "caching": "ReadWrite",
+                        "managedDisk": {
+                            "storageAccountType": "Premium_LRS"
```
can this be SSD for the pinned VM?

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/prereqs/prereq.azuredeploy.json:
```diff
@@ -0,0 +1,269 @@
+                        {
+                            "id": "[resourceId('Microsoft.Network/networkInterfaces', parameters('networkInterfaces_avgvm980_name'))]",
+                            "properties": {
+                                "deleteOption": "Detach"
```
change this to delete

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/prereqs/prereq.azuredeploy.json:
```diff
@@ -0,0 +1,269 @@
+            }
+        }
+    ],
+    "outputs": {
```
vnet name is requested by the main template but is never output

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/prereqs/prereq.azuredeploy.parameters.json:
```diff
@@ -0,0 +1,12 @@
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+      "adminUsername": {
+        "value": "quickstarts"
```
use GEN-UNIQUE

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/README.md:
```diff
@@ -0,0 +1,129 @@
+
+Note: For Data backup and Log backup volumes, user will have an option to choose between nfsv3 or nfsv41. However for Data, Log and Shared volume default will be nfsv41.
+
+## Deployment steps
```
remove this section, it's the same for all samples in the repo

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/prereqs/prereq.azuredeploy.json:
```diff
@@ -0,0 +1,269 @@
+{
```
readme says we need an avset for pinning, but there is no avset in this template

abhishekkrm replied:
avset is required for manual pinning, more details in email thread.

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/azuredeploy.parameters.json:
```diff
@@ -0,0 +1,48 @@
+      "value": " "
+    },
+    "VirtualNetwork": {
+      "value": "GET-PREREQ-virtualNetworkName"
```
this output does not exist

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/azuredeploy.json:
```diff
@@ -0,0 +1,618 @@
+        "ProximityPlacementGroupName": {
+            "type": "string",
+            "minLength": 1,
+            "maxLength": 1024,
```
use correct maxLength values for all parameters

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/azuredeploy.json:
```diff
@@ -0,0 +1,618 @@
+        },
+        "DelegatedSubnet": {
+            "type": "string",
+            "defaultValue": "default",
```
remove defaultValue

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/azuredeploy.json:
```diff
@@ -0,0 +1,618 @@
+            "tags": "[variables('volumeTag')]",
+            "properties": {
+                "creationToken": "[concat(variables('logVolumeNamePrefix'), padLeft(1, variables('paddedLength'), '0'))]",
+                "proximityPlacementGroup": "[resourceId(parameters('ProximityPlacementGroupResourceGroup'),'/providers/Microsoft.Compute/proximityPlacementGroups', parameters('ProximityPlacementGroupName'))]",
```
must be: 

```
[resourceId(parameters('ProximityPlacementGroupResourceGroup'),'Microsoft.Compute/proximityPlacementGroups', parameters('ProximityPlacementGroupName'))]",
```

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/azuredeploy.json:
```diff
@@ -0,0 +1,618 @@
+            "tags": "[variables('volumeTag')]",
+            "properties": {
+                "creationToken": "[concat(variables('dataVolumeNamePrefix'), padLeft(1, variables('paddedLength'), '0'))]",
+                "proximityPlacementGroup": "[resourceId(parameters('ProximityPlacementGroupResourceGroup'),'/providers/Microsoft.Compute/proximityPlacementGroups', parameters('ProximityPlacementGroupName'))]",
```
use:
```
[resourceId(parameters('ProximityPlacementGroupResourceGroup'),'Microsoft.Compute/proximityPlacementGroups', parameters('ProximityPlacementGroupName'))]
```

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/azuredeploy.json:
```diff
@@ -0,0 +1,618 @@
+            "tags": "[variables('volumeTag')]",
+            "properties": {
+                "creationToken": "[concat(variables('sharedVolumeNamePrefix'), padLeft(1, variables('paddedLength'), '0'))]",
+                "proximityPlacementGroup": "[resourceId(parameters('ProximityPlacementGroupResourceGroup'),'/providers/Microsoft.Compute/proximityPlacementGroups', parameters('ProximityPlacementGroupName'))]",
```
```[resourceId(parameters('ProximityPlacementGroupResourceGroup'),'Microsoft.Compute/proximityPlacementGroups', parameters('ProximityPlacementGroupName'))]```

abhishekkrm replied:
done

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/azuredeploy.json:
```diff
@@ -0,0 +1,618 @@
+            "tags": "[variables('volumeTag')]",
+            "properties": {
+                "creationToken": "[variables('dataBackupVolumeNamePrefix')]",
+                "proximityPlacementGroup": "[resourceId(parameters('ProximityPlacementGroupResourceGroup'),'/providers/Microsoft.Compute/proximityPlacementGroups', parameters('ProximityPlacementGroupName'))]",
```
```[resourceId(parameters('ProximityPlacementGroupResourceGroup'),'Microsoft.Compute/proximityPlacementGroups', parameters('ProximityPlacementGroupName'))]```

abhishekkrm replied:
done

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/azuredeploy.json:
```diff
@@ -0,0 +1,618 @@
+            "tags": "[variables('volumeTag')]",
+            "properties": {
+                "creationToken": "[variables('logBackupVolumeNamePrefix')]",
+                "proximityPlacementGroup": "[resourceId(parameters('ProximityPlacementGroupResourceGroup'),'/providers/Microsoft.Compute/proximityPlacementGroups', parameters('ProximityPlacementGroupName'))]",
```
```
[resourceId(parameters('ProximityPlacementGroupResourceGroup'),'Microsoft.Compute/proximityPlacementGroups', parameters('ProximityPlacementGroupName'))]
```

abhishekkrm replied:
done

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/azuredeploy.json:
```diff
@@ -0,0 +1,618 @@
+        {
+            "type": "Microsoft.NetApp/netAppAccounts/volumeGroups",
+            "apiVersion": "2021-06-01",
+            "comments": "Volume Group creation for master Host. This will be deployed only if Starting SAP HostID ==1 i.e master Host",
```
See: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#sort-order-of-properties for the proper sort order of properties

abhishekkrm replied:
done

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/azuredeploy.json:
```diff
@@ -0,0 +1,618 @@
+                        "tags": "[variables('volumeTag')]",
+                        "properties": {
+                            "creationToken": "[concat(variables('logVolumeNamePrefix'), padLeft(add(copyIndex(), 2), variables('paddedLength'), '0'))]",
+                            "proximityPlacementGroup": "[resourceId(parameters('ProximityPlacementGroupResourceGroup'),'/providers/Microsoft.Compute/proximityPlacementGroups', parameters('ProximityPlacementGroupName'))]",
```
```[resourceId(parameters('ProximityPlacementGroupResourceGroup'),'Microsoft.Compute/proximityPlacementGroups', parameters('ProximityPlacementGroupName'))]```

abhishekkrm replied:
done

bmoore-msft reviewed (changes requested):
also see the pipeline results for other changes that need to be made

abhishekkrm commented:
We need an M series VM for the the pipeline to pass.

bmoore-msft commented on quickstarts/microsoft.netapp/anf-sap-hana/anf-sap-hana-storage/prereqs/prereq.azuredeploy.json:
```diff
@@ -0,0 +1,271 @@
+        },
+        "deployVM": {
+            "type": "bool",
+            "defaultValue": "false"
```
this must be a bool, i.e. remove the quotes

```"defaultValue": false```

abhishekkrm replied:
done

bmoore-msft commented:
/azp run
